src/libasr/codegen/wasm_assembler.h

```diff
@@ -51,1198 +59,1387 @@ void emit_leb128_i64(Vec<uint8_t> &code, Allocator &al, int64_t n) {  // for i64
-}
+    // function to emit 64 bit float
+    void emit_f64(Vec<uint8_t> &code, Allocator &al, double x) {
+        emit_ieee754_f64(code, al, x);
+    }
 
-void emit_u32_b32_idx(Vec<uint8_t> &code, Allocator &al, uint32_t idx,
-                      uint32_t section_size) {
-    /*
-    Encodes the integer `i` using LEB128 and adds trailing zeros to always
-    occupy 4 bytes. Stores the int `i` at the index `idx` in `code`.
-    */
-    Vec<uint8_t> num;
-    num.reserve(al, 4);
-    emit_leb128_u32(num, al, section_size);
-    std::vector<uint8_t> num_4b = {0x80, 0x80, 0x80, 0x00};
-    assert(num.size() <= 4);
-    for (uint32_t i = 0; i < num.size(); i++) {
-        num_4b[i] |= num[i];
-    }
-    for (uint32_t i = 0; i < 4u; i++) {
-        code.p[idx + i] = num_4b[i];
+    // function to emit string
+    void emit_str(Vec<uint8_t> &code, Allocator &al, std::string text) {
+        std::vector<uint8_t> text_bytes(text.size());
+        std::memcpy(text_bytes.data(), text.data(), text.size());
+        emit_u32(code, al, text_bytes.size());
+        for (auto &byte : text_bytes) emit_b8(code, al, byte);
     }
-}
 
-// function to fixup length at the given length index
-void fixup_len(Vec<uint8_t> &code, Allocator &al, uint32_t len_idx) {
-    uint32_t section_len = code.size() - len_idx - 4u;
-    emit_u32_b32_idx(code, al, len_idx, section_len);
-}
+    void encode_section(Vec<uint8_t> &des, Vec<uint8_t> &section_content,
+                        Allocator &al, uint32_t section_id,
+                        uint32_t no_of_elements) {
+        // every section in WebAssembly is encoded by adding its section id,
+        // followed by the content size and lastly the contents
+        emit_u32(des, al, section_id);
+        emit_u32(des, al, 4U /* size of no_of_elements */ + section_content.size());
+        uint32_t len_idx = emit_len_placeholder(des, al);
+        emit_u32_b32_idx(des, al, len_idx, no_of_elements);
+        for (auto &byte : section_content) {
+            des.push_back(al, byte);
+        }
+    }
 
-// function to emit length placeholder
-uint32_t emit_len_placeholder(Vec<uint8_t> &code, Allocator &al) {
-    uint32_t len_idx = code.size();
-    code.push_back(al, 0x00);
-    code.push_back(al, 0x00);
-    code.push_back(al, 0x00);
-    code.push_back(al, 0x00);
-    return len_idx;
-}
+    public:
+    Vec<uint8_t> m_type_section;
+    Vec<uint8_t> m_import_section;
+    Vec<uint8_t> m_func_section;
+    Vec<uint8_t> m_memory_section;
+    Vec<uint8_t> m_global_section;
+    Vec<uint8_t> m_export_section;
+    Vec<uint8_t> m_code_section;
+    Vec<uint8_t> m_data_section;
+
+    // no_of_types indicates total (imported + defined) no of functions
+    uint32_t no_of_types;
+    uint32_t no_of_functions;
+    uint32_t no_of_memories;
+    uint32_t no_of_globals;
+    uint32_t no_of_exports;
+    uint32_t no_of_imports;
+    uint32_t no_of_data_segments;
+    uint32_t avail_mem_loc;
+    uint32_t digits_mem_loc;
```

Comment:
I would wish the above variables to be private and not accessible by outer entities.